apps/openmw/mwmechanics/spellcasting.cpp

```diff
@@ -763,6 +763,35 @@ namespace MWMechanics
 
         int school = 0;
 
+        if(mCaster.getClass().isActor()
+                && MWBase::Environment::get().getWorld()->getStore().get<ESM::MagicEffect>()
+                .find(ESM::MagicEffect::WaterWalking))
+        {
+            const MWWorld::CellStore *cell = mCaster.getCell();
+            
+            if(cell->getCell()->hasWater())
+            {
+                const ESM::Position casterPosition = mCaster.getRefData().getPosition();
+                const float casterHeight = mCaster.getCellRef().getScale();
+
+                if(MWBase::Environment::get().getWorld()->castRay(
+                        casterPosition.pos[0], casterPosition.pos[1], casterPosition.pos[2],
+                        casterPosition.pos[0], casterPosition.pos[1], cell->getWaterLevel() + casterHeight ))
+                {
+                    // not enough space to place caster
+                    MWBase::Environment::get().getWindowManager()->messageBox ("#{sMagicInvalidEffect}");
+                    return 0;
```

Comment:
Should not return here. The spell being cast might have additional effects other than water walking, which we still need to apply.